five_tips_about_testable_golang_code.md

```diff
@@ -0,0 +1,171 @@
+
+Test a function like that in Python is easy. Define `mock.Aggregation` somewhere, pass this mock in, assert it be called with given parameter, then assert mock been returned. done.
+
+But we cannot do the same thing in golang.
+
+Remember, golang is a strong type language. If a function takes `elastic.SearchService` as the parameter, you cannot pass anything else. The only possible way is to create a real `elastic.SearchService` object, and do some tweaking in `Aggregation` method. It is complex, and when you seeking the solution about "how to create a real elastic.SearchService object", it is easy to get lost in chasing infinite rabbits in infinite holes. You will forget what you want at first. ~~Just like a bad marriage.~~
+
+Good guys have a solution for this. Instead of a struct, passing an interface can solve it.
+Still same function:
+```go
+func addAggregation(search *SearchServiceInterface) *SearchServiceInterface {
+    search.Aggregation(/* put some variable here*/)
+    search.Aggregation(/*put some other variable here*/)
+
+    return search
+}
+
+type SearchServiceInterface interface {
+    Aggregation(/* exactly same variable here*/)
+}
+```
+
+It looks almost the same, the only difference is instead of `elastic.SearchService`, we passed in an `ElasticRepositoryInterface`. For this example, the compiler would not complain about the object passed in if it has `Aggregation(/* correct type variable here*/)` method. that means we can pass a mock like this into it for testing:
+
+```go
+type SearchServiceMock struct {
+    mock.Mock
+}
+(ssm *SearchServiceMock) Aggregation(/* exactly same variable here*/) {
+    args := ssm.Called(/* exactly same variable here*/)
+    return args.Get(0).(*SearchServiceMock)
+}
+```
+Then we can have all benefits of strong type language, and still keep the flexible of testability.
+
+Of course, as a bad guy who wants to make golang code untestable, you will just pass a struct into the function, nailed it.
+
+## 3. write function instead of method
+TL, DR; Function is not mockable.
+
+Let's start with example:
+
+```go
+func (s *scoreRepository) Create(ctx context.Context, accountGroupID string) (int64, error) {
+    listingPointScore, err := calculateListingScore(ctx, accountGroupID)
+
+    /*
+        some logic we want to test
+    */
+
+    return int64(listingPointScore), nil
+}
+```
+
+If same logic happens in Python, testing will be easy. `@patch` the `calculateListingScore` function and let it return `any_number_we_need, nil`, then verify the result.
+
+But in the cruel world of golang, everything is getting harder: we do not have `@patch`, and we cannot mocking function. If you call a function in somewhere, you cannot mock it at all. If you want to test a function/method included an external function call, you have to call it in real. that is not good.
+
```

Comment:
"But in the cruel world of Golang, everything is harder. We do not have '@patch' like we do in Python. If you call a function somewhere, you cannot mock it at all. If you want to test a function/method that includes an external function call, you have to call that function for real. That is not good."